### DISCUSSION TITLE
Send absolute keys not influenced by X11 remapping

### DISCUSSION THREAD

**Discussion by tobiasBora** [kmonad/kmonad#848] [General]
If I have a, say, azerty keyboard, when I press a `a` the processing will go like this (certainly oversimplified):

> I press `a` -> the keyboard sends a `q` to the kernel (azerty is US with simply a different ink on the keys) -> xkb/ibus/x11/wayland will catch the keycode `q` and map it to `a` when configured in the azerty layout.

Now, Kmonad basically emulates a new keyboard, and is therefore injected between the kernel and X11/wayland. So for instance if I define Kmonad with [ISO 100](https://github.com/kmonad/kmonad/blob/master/keymap/template/iso_100.kbd)'s layout and underscores everywhere in the `deflayer` part, it will go like: 

>  I press `a` -> the keyboard sends a `q` to the kernel (azerty is US with simply a different ink on the keys) -> Kmonad maps `q` to `q` -> xkb/ibus/x11/wayland will catch the keycode `q` and map it to `a` when configured in the azerty layout.

This is fine if I want to handle the layout logic via setxkbmap/ibus/… But the problem is that if I want to implement a new layout directly inside kmonad, then the layout will be applied once by kmonad, and a second time by X11/wayland, resulting in an impredictible key that depends on the layout configured in X11/wayland.

It would be great to provide a way to send absolute key values, that are not further processed by X11/wayland. I'm not sure how realistic this is, but for now I can think of the following options:
1. A first easy (but not very robust) solution would be to at least provide a platform-agnostic way for the user to reset the X11/wayland layout to the `US` layout. For now I do it via:
```
(defcfg
  ;; Get with find /dev/input/ | rg kbd
  input  (device-file "/dev/input/by-path/platform-i8042-serio-0-event-kbd")
  output (uinput-sink "My KMonad output"
    ;; To understand the importance of the following line, see the section on
    ;; Compose-key sequences at the near-bottom of this file.
    "/run/current-system/sw/bin/sleep 1 && /run/current-system/sw/bin/setxkbmap -option compose:menu us")
  cmp-seq compose
  cmp-seq-delay 5 ;; 5ms delay between each compose-key sequence press
  fallthrough true
  allow-cmd true
)
```
but it only works on X11, and if the user later changes the layout manually, it will completely mess-up the system.
2. a second solution would be to create two keyboards, one "relative" one (as for now) that is translated by ibus/xkb/… and one "absolute" one, and somehow (how?) find a way to disable XKB/ibus on this specific keyboard. I found that `setxkbmap` can be called with `setxkbmap -device 11` to only apply to a single keyboard, but gnome/… will certainly apply `setxkbmap` on all keyboards so it would be cool to find a way to avoid this.
3. If it is not possible to do that, then we can maybe generate directly the absolute events from kmonad (or a separate client), getting inspired by how `xdotool` works (I can type `xdotool type "hello"` and, at least with X11, it generates the good keys irrespective of the current layout), or by getting inspired by how virtual keyboards work (for instance using libei that is now supported in gnome wayland)
4. A last solution would be to create again 2 keyboards, then use `setxkbmap -device XX` or alike to set one of the keyboard to US, and then to periodically check (like one per second, not sure if we can do this automatically) if the layout has been changed on this second device. If so, we either can either reset it to us, or just kill kmonad, or reset the layout of kmonad to the default one.

**Comment by tobiasBora** [comment]
I'm actually thinking that in the meantime, it might be easier to:
1. Force the layout to be US
2. Use https://stackoverflow.com/questions/35569562/how-to-catch-keyboard-layout-change-event-and-get-current-new-keyboard-layout-on to detect any layout change by the user (maybe an equivalent exists on wayland), or simply by calling `setxkbmap -query` every second or so
3. On a layout change, kill kmonad (it means that the user wants to use their own layout)

**Comment by slotThe** [comment]
> This is fine if I want to handle the layout logic via setxkbmap/ibus/… But the problem is that if I want to implement a new layout directly inside kmonad, then the layout will be applied once by kmonad, and a second time by X11/wayland, resulting in an impredictible key that depends on the layout configured in X11/wayland.

The default on X11 (and I'm sure on Wayland too) is and will probably always be `en_US`, so I'm not sure how unpredictable this really is. If you set your environment to always apply a certain localisation to all connected input devices, this seems like a conscious choice more than "unpredictable behaviour". (and if it is unpredictable, complain to whatever program is doing that :))

Tools like `xdotool` work a level higher than kmonad (in particular, `xdotool` gets X11 Keysyms as input and not kernel-level keycodes!) and are not really comparable.

> I'm actually thinking that in the meantime, it might be easier to:
>
> […]

This seems fine, but should be done on at the config-level; there are many people who do use X11 for localisation, despite using kmonad for their layout.

**Comment by tobiasBora** [comment]
Thanks for your answer.

> The default on X11 (and I'm sure on Wayland too) is and will probably always be en_US,

Well when I install a system, I am asked which layout I want, so in a sense it depends on the system. For instance on NixOs I use for instance:
```
  services.xserver.xkb.layout = "fr";
```
to default to `fr` for all users. Worse, if I give my machine to a friend, they may type `setxkbmap de` and they would certainly be very confused to see that the layout is completely broken. Similarly, if a user decides to use ibus Win+space to change the layout, they would also completly break the layout of the user. Same with KDE/Gnome default layout system. My goal is, somehow, to prevent this from happening.

Would you at least consider documenting the fact that Kmonad will be later changed by xkb/ibus/… and that one must ensure to use `en_US` for most layout-related options to work?

> Tools like xdotool work a level higher than kmonad (in particular, xdotool gets X11 Keysyms as input and not kernel-level keycodes!) and are not really comparable.

Sure, this is just to illustrate that this is theoretically possible. I was hoping that some libinput tricks would allow kmonad to send an absolute string to print instead of a keycode… seems like it is not possible for what I can tell…

> This seems fine, but should be done on at the config-level; there are many people who do use X11 for localisation, despite using kmonad for their layout.

Sure, I don't want to run this automatically by default. But since this is non trivial to code (especially in a portable way), this might be a simple configuration option that one could provide to the end user.
